File: mcp-server/src/tools/getCashflowHistory.ts
import { z } from 'zod';
import { apiGet } from '../utils/apiClient.js';

export const getCashflowHistoryTool = {
  name: 'get_cashflow_history' as const,
  description: 'Get cashflow data over time, grouped by week, month, or year. Returns income, expenses, and net for each period.',
  parameters: z.object({
    user_id: z.number().default(1).describe('The user ID'),
    granularity: z.enum(['week', 'month', 'year']).default('month').describe('Time period grouping'),
    start_date: z.string().optional().describe('Start date (YYYY-MM-DD)'),
    end_date: z.string().optional().describe('End date (YYYY-MM-DD)'),
  }),
  execute: async (args: { user_id: number; granularity: string; start_date?: string; end_date?: string }) => {
    const params: Record<string, string> = {
      user_id: String(args.user_id),
      granularity: args.granularity,
    };
    if (args.start_date) params.start_date = args.start_date;
    if (args.end_date) params.end_date = args.end_date;

    const data = await apiGet('/api/cashflow/history', params);
    return JSON.stringify(data);
  },
};

File: mcp-server/src/tools/get_cashflow_history.py
import json
from typing import Literal

from fastmcp import FastMCP

from ..utils.api_client import api_get
from ..utils.auth import get_current_token


def register(mcp: FastMCP) -> None:
    @mcp.tool(
        description="Get cashflow data over time, grouped by week, month, or year. "
        "Returns income, expenses, and net for each period.",
    )
    async def get_cashflow_history(
        user_id: int = 1,
        granularity: Literal["week", "month", "year"] = "month",
        start_date: str | None = None,
        end_date: str | None = None,
    ) -> str:
        token = get_current_token()
        params: dict[str, str] = {
            "user_id": str(user_id),
            "granularity": granularity,
        }
        if start_date:
            params["start_date"] = start_date
        if end_date:
            params["end_date"] = end_date

        data = await api_get("/api/cashflow/history", params, token)
        return json.dumps(data)
